packages/eth-providers/src/base-provider.ts

```diff
@@ -347,52 +364,101 @@ export abstract class BaseProvider extends AbstractProvider {
-    );
-  };
+  startSubscriptions = async () => {
+    this.head$ = this.api.rx.rpc.chain.subscribeNewHeads();
+    this.finalizedHead$ = this.api.rx.rpc.chain.subscribeFinalizedHeads();
+
+    const headSub = this.head$.subscribe(header => {
+      this.best$.next({ hash: header.hash.toHex(), number: header.number.toNumber() });
+    });
 
-  _subscribeEventListeners = async () => {
-    const subscriptionMethod = this.safeMode
-      ? this.api.rpc.chain.subscribeFinalizedHeads.bind(this)
-      : this.api.rpc.chain.subscribeNewHeads.bind(this);
+    const finalizedSub = this.finalizedHead$.subscribe(header => {
+      this.finalizedBlockHashes.add(header.hash.toHex());
+      this.finalized$.next({ hash: header.hash.toHex(), number: header.number.toNumber() });
+    });
 
-    const headsUnsub = await subscriptionMethod(async (header: Header) => {
-      // update block cache
-      const blockNumber = header.number.toNumber();
-      const blockHash = header.hash.toHex();
+    await firstValueFrom(this.head$);
+    await firstValueFrom(this.finalizedHead$);
 
-      if (blockNumber === 0) return; // getAllReceiptsAtBlock doesn't work for block 0
-      const receipts = await getAllReceiptsAtBlock(this.api, blockHash);
-      this.blockCache.addReceipts(blockHash, receipts);
+    const safeHead$ = this.safeMode
+      ? this.finalizedHead$
+      : this.head$;
 
-      // eth_subscribe
-      await this._notifySubscribers(blockHash, receipts);
+    const headTasksSub = safeHead$.pipe(
+      // no reciepts for genesis block
+      filter(header => header.number.toNumber() > 0)
+    ).subscribe(header => {
+      const task = this.#async.schedule(this._onNewHead, 0, [header, 5]);
+      this.#headTasks.set(header.hash.toHex(), task);
     });
 
-    const finalizedHeads = this.api.rpc.chain.subscribeFinalizedHeads.bind(this);
-    const finalizedUnsub = await finalizedHeads(async (header: Header) => {
-      if (header.number.toNumber() === 0) return;
-
+    const finalizedTasksSub = this.finalizedHead$.pipe(
+      filter(header => header.number.toNumber() > 0)
+    ).subscribe(header => {
       // notify subscribers
-      const block = await this.getBlockData(header.hash.toHex(), false);
-      const response = hexlifyRpcResult(block);
-      this.eventListeners[SubscriptionType.NewFinalizedHeads].forEach((l) => l.cb(response));
+      const task = this.#async.schedule(this._onNewFinalizedHead, 0, [header, 5]);
+      this.#finalizedHeadTasks.set(header.hash.toHex(), task);
     });
 
     return () => {
-      headsUnsub();
-      finalizedUnsub();
+      headSub.unsubscribe();
+      finalizedSub.unsubscribe();
+      headTasksSub.unsubscribe();
+      finalizedTasksSub.unsubscribe();
     };
   };
 
-  _notifySubscribers = async (blockHash: string, receipts: FullReceipt[]) => {
+  _onNewHead = async ([header, attempts]: [Header, number]) => {
+    attempts--;
+    const blockHash = header.hash.toHex();
+    try {
```

Comment:
curious why we add try catch block here, is there new timing stuff that might make it throw?

Reply:
previously we just called the method and ignored the error. This is implementing a retry with a delay of 1sec